e2e/tests/ganache.rs

```diff
@@ -0,0 +1,106 @@
+use ethcontract::futures::FutureExt;
+use ethcontract::{Http, U256};
+use lazy_static::lazy_static;
+use shared::{transport::LoggingTransport, Web3};
+use std::{
+    fmt::Debug,
+    future::Future,
+    panic::{self, AssertUnwindSafe},
+    sync::Mutex,
+};
+use web3::{api::Namespace, helpers::CallFuture, Transport};
+
+lazy_static! {
+    static ref GANACHE_MUTEX: Mutex<()> = Mutex::new(());
+}
+
+const NODE_HOST: &str = "http://127.0.0.1:8545";
+
+/// *Testing* function that takes a closure and executes it on Ganache.
+/// Before each test, it creates a snapshot of the current state of the chain.
+/// The saved state is restored at the end of the test.
+///
+/// Note that tests calling with this function will not be run aymultaneously.
+pub async fn test<F, Fut>(f: F)
+where
+    F: FnOnce(Web3) -> Fut,
+    Fut: Future<Output = ()>,
+{
+    // The mutex guarantees that no more than a test at a time is running on
+    // Ganache.
+    // Note that the mutex is expected to become poisoned if a test panics. This
```

Comment:
Wouldn't the `panic` get caught with the `catch_unwind()` :point_down: and then have the `Mutex` not get poisoned? Either way, I like not caring about poisoned mutex here since [`catch_unwind` doesn't work for all panics](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html):

> Additionally, this function is not guaranteed to catch all panics

Reply:
By trying it out it looks like the mutex becomes poisoned in all tests after the first one failing, I wasn't able to find a way to "unpoison" a mutex manually.